src/components/Users/CreateUserForm.tsx

```diff
@@ -92,9 +100,131 @@ interface Props {
   onSubmitSuccess?: (user: UserBase) => void;
 }
 
+export const validateRule = (
+  condition: boolean,
+  content: JSX.Element | string,
+  isInitialState: boolean = false,
+) => {
+  return (
+    <div>
+      {isInitialState ? (
+        <CareIcon icon="l-circle" className="text-xl text-gray-500" />
+      ) : condition ? (
+        <CareIcon icon="l-check-circle" className="text-xl text-primary-500" />
+      ) : (
+        <CareIcon icon="l-times-circle" className="text-xl text-red-500" />
+      )}{" "}
+      <span
+        className={classNames(
+          isInitialState
+            ? "text-black"
+            : condition
+              ? "text-primary-500"
+              : "text-red-500",
+        )}
+      >
+        {content}
+      </span>
+    </div>
+  );
+};
+
 export default function CreateUserForm({ onSubmitSuccess }: Props) {
   const { t } = useTranslation();
 
+  const userExistsEnums = {
+    idle: 0,
+    checking: 1,
+    exists: 2,
+    available: 3,
+  };
+
+  const [usernameExists, setUsernameExists] = useState<number>(
+    userExistsEnums.idle,
+  );
+  const [usernameInput, setUsernameInput] = useState("");
+
+  const { data, error, isLoading } = useQuery({
+    queryKey: ["checkUsername", usernameInput],
+    queryFn: query(userApi.checkUsername, {
+      pathParams: { username: usernameInput },
+      silent: true,
+    }),
+    enabled: validateUsername(usernameInput),
+  });
+
+  useEffect(() => {
+    if (!validateUsername(usernameInput)) {
+      setUsernameExists(userExistsEnums.idle);
+      return;
+    }
+
+    if (isLoading) {
+      setUsernameExists(userExistsEnums.checking);
+    } else if (error) {
+      if (error instanceof Error && "status" in error) {
+        const status = (error as any).status;
+
+        if (status === 409) {
+          setUsernameExists(userExistsEnums.exists);
+        } else if (status === 404) {
+          toast.error(
+            "Some error occurred while checking username availability. Please try again later.",
+          );
+          setUsernameExists(userExistsEnums.idle);
+        }
+      } else {
+        toast.error(
+          "Some error occurred while checking username availability. Please try again later.",
+        );
+        setUsernameExists(userExistsEnums.idle);
+      }
+    } else {
+      setUsernameExists(userExistsEnums.available);
+    }
+  }, [data, error, isLoading, usernameInput]);
```

Comment:
i don't see any reason for complicating a simple username check to this level. can't we simply do a useQuery and check if data (available) or error (not available) is present and isLoading (processing stage)?